### ISSUE TITLE
collapse vendor folder in diffs

### ISSUE THREAD

**Issue by nathany** [github/linguist#3234]
The `vendor/` folder is used by Go 1.6+ and other tools/languages for third-party code. While it may be good to review all the code being used in a production app, the GitHub UI still doesn't handle large amounts of code that well. 

> Sorry, we could not display the entire diff because it was too big.

(It's possible to pick commits or files, but the initial load is still an issue)

Happy to open a pull request modifying `generated.rb` to hide vendor/** in diffs.
ref: https://github.com/github/linguist/pull/1549

**Comment by pchaigno** [comment]
@arfon Shoudn't github.com hide all vendored files in pull requests by default? That sounds like an issue for upstream...

**Comment by nathany** [comment]
I don't know -- only that linguist is currently responsible for collapsing the Godeps folder -- https://github.com/github/linguist/blob/master/lib/linguist/generated.rb#L314.

Something I had suggested two years ago, before Go standardized around `/vendor/`. 

Ref: https://github.com/tools/godep/issues/121

**Comment by nathany** [comment]
I'd be happy to send a pull request to update `generated.rb` if desired.

**Comment by arfon** [comment]
According to [this line](https://github.com/github/linguist/blob/master/lib/linguist/vendor.yml#L78) we consider this `vendored` which means it's excluded from the language statistics but I htink we need to update `generated.rb` as @nathany suggests to have them excluded from diffs.

> I'd be happy to send a pull request to update `generated.rb` if desired.

@nathany that would be great thanks!

**Comment by nathany** [comment]
Will do.

I'm only concerned with the `vendor/` folder as used by Go tools. Would you like `third-party`, `extern`, etc. collapsed in diffs as well?

**Comment by pchaigno** [comment]
@arfon Does this mean we should treat every vendored files as generated? Wouldn't that be easier to handle upstream?

**Comment by arfon** [comment]
> Wouldn't that be easier to handle upstream?

How do you mean sorry?

**Comment by pchaigno** [comment]
Sorry that was unclear. I guess there is some closed source code that checks whether a file is `generated` before choosing to collapse it in diffs. Maybe this same code could also check for `vendored` files...?

**Comment by nathany** [comment]
@pchaigno Obviously I'm not familiar with any closed source you may prefer to change instead. I'm just basing this on the previous change for Godeps. 

The pull request is up at https://github.com/github/linguist/pull/3262. 
Please let me know you would like any changes.

**Comment by arfon** [comment]
> Sorry that was unclear. I guess there is some closed source code that checks whether a file is `generated` before choosing to collapse it in diffs. Maybe this same code could also check for `vendored` files...?

Ahh, gotcha. So we currently mixin [`BlobHelper`](https://github.com/github/linguist/blob/master/lib/linguist/blob_helper.rb) into the view class that renders the diff on GitHub so we literally do something like:

```
<% unless diff_blob.generated? %>
  RENDER DIFF CONTENTS HERE
<% end %>
```

We could certainly expand that to include `vendored?` too:

```
<% unless diff_blob.generated? || diff_blob.vendored? %>
  RENDER DIFF CONTENTS HERE
<% end %>
```

which is what I think you're suggesting @pchaigno?

**Comment by nathany** [comment]
Is the suggestion to make a clear distinction between vendored and generated? For example, `node_modules` is vendored code, not generated. Would the distinction be useful in the UI? I kind've feel like that is a larger refactoring that should be a separate change from this.

Personally, I consider #3262 a temporary fix. Taking a look at `vendor/` code and generated code is considered a good practice by some -- myself included. For now, I think we'll continue to vendor code in a separate commit so that it can be reviewed if desired, separately from the newly written code that uses it.

This is really to address the sluggish UI when GitHub diffs are too big -- or partial diffs when it knows the diff is too big, but not necessarily what to exclude. For anyone using the `vendor/` folder, the first click on "Files changed" should come up faster once #3262 is in production.

**Comment by pchaigno** [comment]
@arfon Yes, that's it! 

> Is the suggestion to make a clear distinction between vendored and generated? For example, node_modules is vendored code, not generated (https://github.com/github/linguist/blob/master/lib/linguist/generated.rb#L58). 

Exactly! Same for `vendors/`.
I'm :+1: for #3262 being a temporary fix, but I think a proper long term fix would be make the change @arfon wrote for the view class.

**Comment by arfon** [comment]
> I'm 👍 for #3262 being a temporary fix, but I think a proper long term fix would be make the change @arfon wrote for the view class.

I've got an open PR on `github/github` to rework the view class in this way. Assuming I can convince my colleagues that suppressing `vendored?` files in views by default is a good idea then this PR will become unnecessary.

**Comment by nathany** [comment]
Thanks Arfon.

It would be nice to collapse `vendored?` initially when clicking "Files changed" (full diff). That would make the UI more responsive.

Personally, I thin it would be nice to expand vendored/generated code when looking at an individual commit that would otherwise be entirely collapsed -- or have an easy way to expand all. 

If the vendored code is in a separate commit, it can still be reviewed easily, but without slowing down the UI by including all vendored code with the full diff.

**Comment by nathany** [comment]
@arfon Any updates on this? It's incredibly painful to review pull requests with 100s of vendor files.

**Comment by nathany** [comment]
Just a reminder that it's been 20 days since I submitted a pull request to resolve this issue, even if only temporarily. #3262

**Comment by arfon** [comment]
> Just a reminder that it's been 20 days since I submitted a pull request to resolve this issue, even if only temporarily. #3262

Sorry for the very slow response here @nathany. I've recently tested out the more 'complete' fix described [above](https://github.com/github/linguist/issues/3234#issuecomment-251503691) and am waiting on a 👍  to ship this by one of the engineering teams.

If I don't make progress on that today I'll move forward with #3262

**Comment by arfon** [comment]
@nathany - this should be fixed now. Can you confirm? Here's an example PR: https://github.com/chain/chain/commit/a0e993420325dcbb14846d6ffea21d77fd711749

![screen shot 2016-10-25 at 7 07 59 am](https://cloud.githubusercontent.com/assets/4483/19684057/2d276f0e-9a83-11e6-9fad-b7c8a92cc12c.png)

**Comment by nathany** [comment]
Thanks @afron.

Looking at a past pull request, the vendored files are collapsed and I think it loads faster, but I was hoping it would solve an issue that it doesn't appear to.

> Sorry, we could not display the entire diff because it was too big.

<img width="1027" alt="gocql" src="https://cloud.githubusercontent.com/assets/4566/19687748/ee628e60-9a83-11e6-9692-59958bf9e57b.png">

If it's always `vendor/` code at the bottom, then that's not too much of a problem, but I think there is still room for improvement.

This is a step in the right direction though. Thanks very much for making it happen!

**Comment by arfon** [comment]
> If it's always `vendor/` code at the bottom, then that's not too much of a problem, but I think there is still room for improvement.

Understood. 

/ cc @brianmario who has been thinking about how to improve some of the diff views on GitHub recently.

@nathany - are we good to close this issue now?

**Comment by nathany** [comment]
Yup. I trust you are tracking the larger issue with Git diffs internally. 

I don't see any way to improve that from linguist itself.

Cheers.

**Comment by z4y4ts** [comment]
@nathany @arfon sorry for disturbing, but can you please help me with the similar issue for JS/CSS vendored files.

As far as I can see they're not being suppressed in the diffs. By a chance would you know if there plans/PR/issue/workaround to add similar feature for languages other thatn Go?

**Comment by nathany** [comment]
@z4y4ts I'm not sure, but I suggest opening a new issue rather than commenting on a closed pull request.

Fyi, this article describes the improvements made: https://githubengineering.com/how-we-made-diff-pages-3x-faster/
